### DISCUSSION TITLE
Question on When the useAtom Overload Type is Triggered

### DISCUSSION THREAD

**Discussion by theWorkFsz** [pmndrs/jotai#2499] [Q&A] [answered]
Hey guys, I've recently been trying to read and understand the code logic of Jotai. When I came across the overloading of `useAtom`, I had some questions. Under what specific atom definitions would the following overload be triggered?

```ts
//ref: https://github.com/pmndrs/jotai/blob/main/src/react/useAtom.ts#L42
export function useAtom<AtomType extends Atom<unknown>>(
  atom: AtomType,
  options?: Options,
): [Awaited<ExtractAtomValue<AtomType>>, never]
```

Given that its second return value is `never`, it seems to be related to read-only atoms. However, for simpler  readonly atoms like

```ts
const primitiveAtom = atom(0)
const readonlyDerivedAtom1 = atom(() => 0)
const readonlyDerivedAtom2 = atom((get) => get(primitiveAtom))

// Both of the above read-only atoms' corresponding useAtom type should be another overload
// ref: https://github.com/pmndrs/jotai/blob/main/src/react/useAtom.ts#L27
export function useAtom<Value>(
  atom: Atom<Value>,
  options?: Options,
): [Awaited<Value>, never]
```

I've looked through the [unit tests](https://github.com/pmndrs/jotai/blob/main/tests/react/types.test.tsx#L7) and seem to find no test code for this type of overload, so I'm curious about what kind of atom it specifically corresponds to.

**Comment by himself65** [comment]
1. It's [function overloading](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads); the result type depends on function input; in your case, it will be when upper cases do not fit

**Reply by himself65** [reply]
2. I think it should be case [line #27](https://github.com/pmndrs/jotai/blob/d1b74111a761347ed8fe88f1068e6e5f8caf7b10/src/react/useAtom.ts#L27) because the type here is `Atom<number>` which could be infer by typescript

**Comment by dai-shi** [comment]
> I've looked through the [unit tests](https://github.com/pmndrs/jotai/blob/main/tests/react/types.test.tsx#L7) and seem to find no test code for this type of overload

I think if you remove the overload, some tests will fail.

**Reply by theWorkFsz** [reply]
Thank you for your reply.
Following your tips, I tested by separately removing each of the two function overloading and observed:

1. Removing either or both doesn't lead to test failures.
2. Removing either one makes TypeScript fall back to the other overload. If both are removed, TypeScript reports "No overload matches this call." 

Therefore, I guess that only the first function overloading  [line #27](https://github.com/pmndrs/jotai/blob/main/src/react/useAtom.ts#L27)  would be triggered, while the other [line 42](https://github.com/pmndrs/jotai/blob/main/src/react/useAtom.ts#L42 ) is actually redundant.

**Reply by dai-shi** [reply, marked as answer]
Thanks for your investigation. It revealed that some tests are missing. #2503 adds it.
